table

syarikim opened the pull request:
テーブル設計課題
https://tech-essentials.work/courses/7/tasks/8/outputs/1261

syarikim committed:
table

テーブル設計課題

hirohiroiida reviewed (commented):
同じ課題をしている、飯田です！
多分、同じこと、ダイソンさんに言われるかもですが、.md でマークダウンで書けるのでそちらで書く方がプレビューとかもあるのでいいですよ！
僕も、まだ終わってないので頑張りましょー！

syarikim committed:
table課題

syarikim commented:
ご指摘ありがとうございます。

.mdにしてファイル上げました。& 記載も少し見直しました。

DaichiSaito commented on table:
```diff
@@ -0,0 +1,16 @@
+
+Users
+  user_id:int　PK
```
RailsにおいてはPKは`id`ですね。
専門用語で言うとサロゲートキーとかっていいます。

syarikim replied:
ありがとうございます。なるほどです。
ここは、Rails勉強始めてから違和感覚えているところなのですが、下記の理解で合っていますですでしょうか？
特に2点目がピンときてないです。
・Railsの場合、PKは意味を持たないサロゲートキーにすることが一般的。自動設定の「id」にして、tabl定義にも記載しない。
・FKは、参照するtable名+idで定義する。

DaichiSaito replied:
Railsはそういうルールですね！

参考
[Active Record の基礎 \- Railsガイド](https://railsguides.jp/active_record_basics.html#%E3%82%B9%E3%82%AD%E3%83%BC%E3%83%9E%E3%81%AE%E3%83%AB%E3%83%BC%E3%83%AB)

syarikim replied:
ガイドに結構詳しく書いていたのですね。ありがとうございます！

DaichiSaito commented on table:
```diff
@@ -0,0 +1,16 @@
+
+Users
+  user_id:int　PK
+  user_name:string
```
`name`で良いと思います。

DaichiSaito commented on table:
```diff
@@ -0,0 +1,16 @@
+Users
+  user_id:int　PK
+  user_name:string
+  mail:mail
```
動詞っぽく見えるので`email`の方が良いと思いますし型が`mail`っておかしいです。

syarikim replied:
emailにします。mail型はコピペミスりました。
stringに直します。

DaichiSaito commented on table:
```diff
@@ -0,0 +1,16 @@
+  user_id:int　PK
+  user_name:string
+  mail:mail
+  avataara_img:string
```
ActiveStorageを使うならこのカラムは不要ですね。

syarikim replied:
ここは、「railsの教科書」を倣って
carrierwave＋画像情報をDB登録にしようとしてました。

DaichiSaito replied:
であればOKです！

DaichiSaito commented on table:
```diff
@@ -0,0 +1,16 @@
+  mail:mail
+  avataara_img:string
+  admin:boolean default: false
+  password:string
```
passwordを保存するんでしょうか？
ここは現場Railsなどで復習しましょう。

DaichiSaito commented on table:
```diff
@@ -0,0 +1,16 @@
+
+
+Questions
+  questione_id:int PK
```
usersテーブルと同様の指摘です。

DaichiSaito commented on table:
```diff
@@ -0,0 +1,16 @@
+  user_id:int FK
+  title:string 
+  detail:text
+  status:enum
```
`enum`という型はないです。

syarikim replied:
誤解してました。
型ではなく、あくまで状態のことなのですね。
intにします。

syarikim committed:
Delete table

syarikim committed:
up

syarikim commented:
ありがとうございます。
指摘確認しました。

.mdと最初にupしたファイルで２つあったので
最初のファイルを削除しました。.mdの方を確認いただければと思います。

miketa-webprgr commented on table.md:
```diff
@@ -2,30 +2,29 @@ users
+| ID連番         | id      | int     | not_null,uniq                        |
+| 名前           | name    | string  | not_null                             |
+| メールアドレス | email         | string    | uniq                                 |
+| 管理者かどうか | admin_fg     | boolean | not_null,defalut:false                        | 
```
今回はあまり当てはまらないかと思いますが、フラグという命名だと分かりづらくなることがあるので気をつけた方がいいという話を聴いたことがあるので参考までに貼っておきます！　ちょっと強引な感じで説明しますが、adminであることを確認するフラグか、adminが操作できることを禁止するフラグなのか、人によっては解釈が異なってしまうため、場合によっては避けたほうがよいようです！
https://neos21.net/blog/2016/03/08-01.html

あと、flagの省略形としては、`flg`という書き方の方が見ることが多いかなと思いました！

syarikim replied:
なるほどです。
確かに、カラム名はadminでbooleanの方がパッと見分かりやすそうですね。
修正します。

miketa-webprgr replied:
わざわざ修正していただいて、すみません！　ちょっと混乱を招いたかもしれないです🙇 

「flagを使う命名は、場合によっては混乱を招きやすいので気をつけてくださいね」というぐらいの話だったので、今回は別に問題ないと思います！

あと、個人的には、`admin_fg`よりも`admin`の方がシンプルでいいと思います！

syarikim replied:
あ、いえ、僕個人的にも下記で納得したので、今回こちらで行こうと思います！
>個人的には、admin_fgよりもadminの方がシンプルでいいと思います！

miketa-webprgr commented on table.md:
```diff
@@ -2,30 +2,29 @@ users
-| ID連番                | answer_id    | int    | not_null,uniq                        | 
+| ID連番                | id    | int    | not_null,uniq                        | 
 | usersとの紐付けFK     | user_id      | int    | not_null                             | 
 | questionsとの紐付けFK | questions_id | int    | not_null                             | 
```
questions_id? question_id?

syarikim replied:
ありがとうございます。FKは単数系に統一します。

miketa-webprgr replied:
統一するという観点も大事ですが、Railsだとデフォルトで単数系_idを外部キーになります！
（Railsに限らず、多くのフレームワークだとおそらくそう？）

なので、その前提に立つと、単数形にするのが自然だと思います！

syarikim replied:
ありがとうございます。railsガイドにも書いてありました！単数形にします！

miketa-webprgr commented on table.md:
```diff
@@ -2,30 +2,29 @@ users
 | --------------------- | ------------ | ------ | ------------------------------------ | 
-| ID連番                | answer_id    | int    | not_null,uniq                        | 
+| ID連番                | id    | int    | not_null,uniq                        | 
 | usersとの紐付けFK     | user_id      | int    | not_null                             | 
```
users_id? user_id?

syarikim replied:
こちらも単数系に統一するようにします

miketa-webprgr commented on table.md:
```diff
@@ -2,30 +2,29 @@ users
 
 | 属性の意味     | カラム名     | 型      | その他(null、ユニーク、デフォルト値) |
 | -------------- | ------------ | ------- | ------------------------------------ |
```
適当にこのあたりにコメントしますけど、password関係のカラムっていらないんでしょうか？

passwordについては、暗号化しないでそのまま保存することが問題なので、そういった観点からもう一度調べてみるといいと思いますよ！

syarikim replied:
なるほどです。理解しました。
現場railsに合わせて「password_digest」で保存するようにします。

syarikim committed:
Update table.md

修正箇所めも
・各テーブルのPKのxxx_IDを削除しました。
・usersのカラム「avataara_img 」を元に戻しました。(誤って削除したため)
・usersのカラムadmin_fgからadminに修正
・usersにpassword_digest追加
・各テーブルのFKのxxx_IDのxxxを単数形に修正。

syarikim commented:
頂いた指摘を元にテーブル定義見直しました。

修正箇所めも
・各テーブルのPKのxxx_IDを削除しました。
・usersのカラム「avataara_img 」を元に戻しました。(誤って削除したため)
・usersのカラムadmin_fgからadminに修正
・usersにpassword_digest追加
・各テーブルのFKのxxx_IDのxxxを単数形に修正。

DaichiSaito commented on table.md:
```diff
@@ -0,0 +1,30 @@
+| --------------------- | ------------ | ------ | ------------------------------------ | 
+| usersとの紐付けFK     | user_id      | int    | not_null                             | 
+| questionsとの紐付けFK | question_id | int    | not_null                             | 
+| 答えタイトル          | title        | string | not_null                             | 
```
答えのタイトルは不要かなと思います！！
それ直したらマージしちゃいましょう！

syarikim replied:
了解です！
タイトルはモック見ても判断つかなったので、一応記載しただけで
特にこだわりもないので削除します。

syarikim committed:
Update table.md

answersのtitle削除
